experiment/nipopow_verifier/tools/proof/proof.py

```diff
@@ -75,3 +89,216 @@ def set(self, proof, proof_name=""):
+    """
+    Returns a proof's best level and score
+    """
+
+    levels = defaultdict(lambda: 0)
+    for header, _ in proof:
+        level = header_level(header)
+        levels[level] += 1
+
+    top_level = max(levels.keys())
+
+    print("Top level is", top_level)
+    for l in range(top_level):
+        if levels[l] == 0:
+            levels[l] = 0
+
+    keys = sorted(levels.keys(), reverse=True)
+    for i in range(len(keys) - 1):
+        levels[keys[i + 1]] += levels[keys[i]]
+
+    scores = defaultdict(lambda: 0)
+
+    max_score = 0
+    max_level = 0
+    for level in keys:
+        curr_score = levels[level] * pow(2, level)
+        if levels[level] >= miou and curr_score > max_score:
+            max_score = curr_score
+            max_level = level
+        scores[level] = curr_score
+
+    return max_level, max_score, levels, scores
+
+
+def change_interlink_bytes(header, new_interlink):
+    """
+    Changes the header's bytes associated with a blocks's interlink hash.
+    Meaning the first 32
+    """
+
+    if new_interlink == []:
+        return header
+    hashed_interlink = blockchain_utils.hash_interlink(new_interlink)
+    new_header = hashed_interlink + header[32:]
+    return new_header
+
+
+def _isolate_proof_level(proof, level, interlink_map):
+    """
+    Returns a valid proof consisted by blocks of a certain level and above
+    """
+
+    isolated_proof = []
+
+    rev_proof = proof[::-1]
+
+    start_index = None
+
+    # Add the first block of the proof until the first of wanted level
+    for i, block in enumerate(rev_proof):
+        header, _ = block
+        isolated_proof.append(block)
+        if header_level(header) >= level:
+            start_index = i
+            break
+
+    if start_index is None:
+        return []
+
+    for i in range(start_index + 1, len(rev_proof)):
+        if header_level(rev_proof[i][0]) < level:
+            continue
+        header = rev_proof[i][0]
+        header_hash = blockchain_utils.Hash(header)
+        interlink_array = blockchain_utils.list_flatten(interlink_map[header_hash])
+        new_merkle_proof = blockchain_utils.prove_interlink(interlink_array, level)
+        isolated_proof.append((header, new_merkle_proof))
+
+    isolated_proof = isolated_proof[::-1]
+    blockchain_utils.verify_proof(
+        blockchain_utils.Hash(isolated_proof[0][0]), isolated_proof[:-1]
+    )
+
+    return isolated_proof
+
+
+def isolate_proof_level(level, fork_proof, header, header_map, interlink_map):
+    """
+    Returns a valid proof consisted by blocks of a certain level and above
+    """
+
+    start = []
+    for p in fork_proof[::-1]:
+        h, _ = p
+        if header_level(h) >= level:
+            break
+        start.append(p)
+
+    anchor = blockchain_utils.CBlockHeaderPopow.deserialize(h)
+
+    interlink = blockchain_utils.list_flatten(interlink_map[header.GetHash()])
+    proof = []
+    mp = []
+    print("--")
+    i = 0
+    while True:
```

Comment:
Can you rewrite without `while True`? Let's have it be more clear what you're actually iterating over.

Reply:
Done